File: golang/tests/pkg/integration/session_keep_alive_integration_test.go
package integration

import (
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
)

func isReleasedStatusResult(r *agentbay.SessionStatusResult) bool {
	if r == nil {
		return false
	}
	if r.Success {
		return r.Status == "FINISH" || r.Status == "DELETING" || r.Status == "DELETED"
	}
	return r.Code == "InvalidMcpSession.NotFound"
}

// TestSessionKeepAliveIntegration verifies that keep-alive refreshes idle timer:
// after calling KeepAlive on one session, it should remain alive when the control session is released.
func TestSessionKeepAliveIntegration(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Fatal("AGENTBAY_API_KEY environment variable is not set")
	}

	client, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to initialize AgentBay client: %v", err)
	}

	idleReleaseTimeoutSeconds := int32(30)
	maxOverSeconds := int32(60)
	pollInterval := 15 * time.Second
	imageID := "linux_latest"

	commonLabels := map[string]string{
		"test": "session-keep-alive",
		"sdk":  "golang",
	}

	controlParams := agentbay.NewCreateSessionParams().
		WithImageId(imageID).
		WithIdleReleaseTimeout(idleReleaseTimeoutSeconds).
		WithLabels(map[string]string{
			"test": commonLabels["test"],
			"sdk":  commonLabels["sdk"],
			"role": "control",
		})
	refreshedParams := agentbay.NewCreateSessionParams().
		WithImageId(imageID).
		WithIdleReleaseTimeout(idleReleaseTimeoutSeconds).
		WithLabels(map[string]string{
			"test": commonLabels["test"],
			"sdk":  commonLabels["sdk"],
			"role": "refreshed",
		})

	start := time.Now()

	controlCreate, err := client.Create(controlParams)
	if err != nil {
		t.Fatalf("Failed to create control session: %v", err)
	}
	if controlCreate == nil || controlCreate.Session == nil {
		t.Fatalf("Create returned nil control session")
	}
	control := controlCreate.Session

	refreshedCreate, err := client.Create(refreshedParams)
	if err != nil {
		t.Fatalf("Failed to create refreshed session: %v", err)
	}
	if refreshedCreate == nil || refreshedCreate.Session == nil {
		t.Fatalf("Create returned nil refreshed session")
	}
	refreshed := refreshedCreate.Session

	defer func() {
		for _, s := range []*agentbay.Session{refreshed, control} {
			if s == nil {
				continue
			}
			status, e := s.GetStatus()
			if e == nil && status != nil && !isReleasedStatusResult(status) {
				_, _ = s.Delete()
			}
		}
	}()

	time.Sleep(time.Duration(idleReleaseTimeoutSeconds) * time.Second / 2)
	keepAliveResult, err := refreshed.KeepAlive()
	if err != nil {
		t.Fatalf("KeepAlive failed: %v", err)
	}
	if keepAliveResult == nil || !keepAliveResult.Success {
		t.Fatalf("KeepAlive returned success=false: %v", keepAliveResult)
	}

	deadline := start.Add(time.Duration(idleReleaseTimeoutSeconds+maxOverSeconds) * time.Second)
	for time.Now().Before(deadline) {
		controlStatus, e1 := control.GetStatus()
		if e1 != nil {
			t.Fatalf("Control GetStatus failed: %v", e1)
		}
		refreshedStatus, e2 := refreshed.GetStatus()
		if e2 != nil {
			t.Fatalf("Refreshed GetStatus failed: %v", e2)
		}

		if isReleasedStatusResult(controlStatus) {
			if isReleasedStatusResult(refreshedStatus) {
				t.Fatalf("Refreshed session released no later than control session; keep-alive did not extend idle timer")
			}
			return
		}

		time.Sleep(pollInterval)
	}
}

File: typescript/tests/integration/session-keep-alive.integration.test.ts
import { AgentBay } from "../../src";
import { getTestApiKey } from "../utils/test-helpers";
import { log } from "../../src/utils/logger";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isReleased(status: any): boolean {
  if (!status) {
    return false;
  }
  if (
    status.success === false &&
    status.code === "InvalidMcpSession.NotFound"
  ) {
    return true;
  }
  if (
    status.success === true &&
    ["FINISH", "DELETING", "DELETED"].includes(status.status)
  ) {
    log(`✅ Session released: ${status.status}`);
    return true;
  }
  return false;
}

describe("Session keepAlive (integration)", () => {
  it(
    "should keep refreshed session alive longer than control session",
    async () => {
      const apiKey = getTestApiKey();
      const agentBay = new AgentBay({ apiKey });

      const idleReleaseTimeoutSeconds = 30;
      const maxOverSeconds = 60;
      const pollIntervalMs = 2000;
      const imageId = "linux_latest";

      const start = Date.now();

      const commonLabels = {
        test: "session-keep-alive",
        sdk: "typescript",
      };

      const controlCreate = await agentBay.create({
        imageId,
        idleReleaseTimeout: idleReleaseTimeoutSeconds,
        labels: { ...commonLabels, role: "control" },
      });
      expect(controlCreate.success).toBe(true);
      expect(controlCreate.session).toBeDefined();
      const control = controlCreate.session!;

      const refreshedCreate = await agentBay.create({
        imageId,
        idleReleaseTimeout: idleReleaseTimeoutSeconds,
        labels: { ...commonLabels, role: "refreshed" },
      });
      expect(refreshedCreate.success).toBe(true);
      expect(refreshedCreate.session).toBeDefined();
      const refreshed = refreshedCreate.session!;

      try {
        await sleep((idleReleaseTimeoutSeconds * 1000) / 2);
        const keepAliveResult = await refreshed.keepAlive();
        expect(keepAliveResult.success).toBe(true);

        const deadline =
          start + (idleReleaseTimeoutSeconds + maxOverSeconds) * 1000;
        while (Date.now() < deadline) {
          const controlStatus = await control.getStatus();
          const refreshedStatus = await refreshed.getStatus();

          if (isReleased(controlStatus)) {
            expect(isReleased(refreshedStatus)).toBe(false);
            const elapsed = (Date.now() - start) / 1000;
            log(
              `✅ Control released while refreshed alive, elapsed=${elapsed.toFixed(
                2
              )}s`
            );
            return;
          }

          // Check if refreshed session was released before control session (unexpected)
          if (isReleased(refreshedStatus)) {
            throw new Error(
              "Refreshed session was released before control session; keep-alive may have failed"
            );
          }

          await sleep(pollIntervalMs);
        }
      } finally {
        // Best-effort cleanup: delete both if still alive
        for (const s of [refreshed, control]) {
          try {
            const status = await s.getStatus();
            if (!isReleased(status)) {
              await s.delete();
            }
          } catch (e) {
            // ignore
          }
        }
      }
    },
    2 * 60 * 1000
  );
});
